File: 06.js
const fs = require('fs')
const readline = require('readline')
const events = require('events')
const input_file = "06input.txt"
const data = fs.readFileSync(input_file, {encoding:'utf8', flag:'r'});

function main(pck_len, file) {
    const str = data.split("\n")[0]
    const result = str
        .split("")
        .reduce((a, b, i, arr) => {
            if (a.length === pck_len) {
                if (new Set(a).size === pck_len) {
                    arr.splice(0)
                    return a
                }
                else {
                    a.shift()
                }
            }
            a.push(b)
            return a
        }, [])
        .join("")
    console.log(str.indexOf(result)+pck_len)

}

async function doit() {
    await main(4, input_file)
    await main(14, input_file)
}

doit()

File: 06.py
import os
import sys
import functools

curr_dir = os.path.dirname(os.path.realpath(__file__))
input_file = f"{curr_dir}/06input.txt"

def find_sequence(a, b, length):
    if len(a) == length:
        if len(set(a)) == length:
            return a
        else:
            a = a[1:]
    a += b
    return a

def main(pck_len, file):
    with open(file, "r") as f:
        str = "".join(f).split("\n")[0]
        result = functools.reduce(lambda a,b: find_sequence(a,b,pck_len), list(str))
        print(str.find(result)+pck_len)

main(4, input_file)
main(14, input_file)


